plugins/dummy_sha256/src/hipcheck_transport.rs

```diff
@@ -0,0 +1,226 @@
+use crate::hipcheck::{Query as PluginQuery, QueryState};
+use anyhow::{anyhow, Result};
+use indexmap::map::IndexMap;
+use serde_json::Value;
+use std::collections::VecDeque;
+use std::sync::Arc;
+use tokio::sync::{mpsc, Mutex};
+use tonic::{codec::Streaming, Status};
+
+#[derive(Debug)]
+pub struct Query {
+	pub id: usize,
+	// if false, response
+	pub request: bool,
+	pub publisher: String,
+	pub plugin: String,
+	pub query: String,
+	pub key: Value,
+	pub output: Value,
+}
+impl TryFrom<PluginQuery> for Query {
+	type Error = anyhow::Error;
+	fn try_from(value: PluginQuery) -> Result<Query> {
+		use QueryState::*;
+		let request = match TryInto::<QueryState>::try_into(value.state)? {
+			QueryUnspecified => return Err(anyhow!("unspecified error from plugin")),
+			QueryReplyInProgress => {
+				return Err(anyhow!(
+					"invalid state QueryReplyInProgress for conversion to Query"
+				))
+			}
+			QueryReplyComplete => false,
+			QuerySubmit => true,
+		};
+		let key: Value = serde_json::from_str(value.key.as_str())?;
+		let output: Value = serde_json::from_str(value.output.as_str())?;
+		Ok(Query {
+			id: value.id as usize,
+			request,
+			publisher: value.publisher_name,
+			plugin: value.plugin_name,
+			query: value.query_name,
+			key,
+			output,
+		})
+	}
+}
+impl TryFrom<Query> for PluginQuery {
+	type Error = anyhow::Error;
+	fn try_from(value: Query) -> Result<PluginQuery> {
+		let state_enum = match value.request {
+			true => QueryState::QuerySubmit,
+			false => QueryState::QueryReplyComplete,
+		};
+		let key = serde_json::to_string(&value.key)?;
+		let output = serde_json::to_string(&value.output)?;
+		Ok(PluginQuery {
+			id: value.id as i32,
+			state: state_enum as i32,
+			publisher_name: value.publisher,
+			plugin_name: value.plugin,
+			query_name: value.query,
+			key,
+			output,
+		})
+	}
+}
+
+#[derive(Clone, Debug)]
+pub struct HcTransport {
+	tx: mpsc::Sender<Result<PluginQuery, Status>>,
+	rx: Arc<Mutex<MultiplexedQueryReceiver>>,
+}
+impl HcTransport {
+	pub fn new(rx: Streaming<PluginQuery>, tx: mpsc::Sender<Result<PluginQuery, Status>>) -> Self {
+		HcTransport {
+			rx: Arc::new(Mutex::new(MultiplexedQueryReceiver::new(rx))),
+			tx,
+		}
+	}
+	pub async fn send(&self, query: Query) -> Result<()> {
```

Comment:
This is just a style thing, but I think in general having empty newlines between functions and in many other places would 1) match the style of the rest of the Hipcheck codebase and 2) be easier to read.

Reply:
I can see that, but unless we make HcSessionSocket itself spawn a task to watch this "drop" message, the best we can do is have HcSessionSocket look for any such messages after `let Some(raw) = rx_handle.message().await?` in `listen()`. Which do you prefer? I personally don't stylistically like the idea of spawning a task inside HcSessionSocket, but up to you

Reply:
I've pushed an example of how it would work without spawning an internal listener task. Let me know what you think.

Reply:
Whoops, I realized I've been replying to the wrong comment. Sorry about that. I'm referring to the above session tracker discussion.

Reply:
Thanks, this does seem cleaner to me.